README.md
# ![WebPary logo](logo.png)

> Fast and light NodeJS library and cli to scrape and interact with websites using [Puppeteer](https://github.com/GoogleChrome/puppeteer) ([or not](#goto)) and [YAML definitions](https://en.wikipedia.org/wiki/YAML)

```yaml
version: 1
jobs:
  main:
    steps:
      - goto: https://github.com/marketplace?category=code-quality
      - pdf:
          path: Github_Tools.pdf
          format: A4
      - many: 
          as: github_tools
          event: githubTool
          selector: main .col-lg-9.mt-1.mb-4.float-lg-right a.col-md-6.mb-4.d-flex.no-underline
          element:
            - property:
                selector: a
                type: string
                property: href
                as: url
                transform: absoluteUrl
            - text:
                selector: h3.h4
                type: string
                transform: trim
                as: name
            - text:
                selector: p
                type: string
                transform: trim
                as: description
```

_Return an array with Github's tools, and creates a PDF. Example output:_

```json
{
  "github_tools": [
    {
      "url": "https://github.com/marketplace/codelingo",
      "name": "codelingo",
      "description": "Your Code, Your Rules - Automate code reviews with your own best practices"
    },
    {
      "url": "https://github.com/marketplace/codebeat",
      "name": "codebeat",
      "description": "Code review expert on demand. Automated for mobile and web"
    },
    ...
  ]
}
```

---

Don't panic. There are examples for all WebParsy features in the examples folder. This are as basic as possible to help you get started.

<div>
	<p>
    <sup>Support my work,</sup>
		<br>
		<a href="https://www.patreon.com/joseconstela">
			<img src="https://c5.patreon.com/external/logo/become_a_patron_button@2x.png" width="160">
		</a>
  </p>
</div>

##### Table of Contents

- [Overview](#overview)
- [Browser config](#browser-config)
- [Output](#output)
- [Transform](#transform)
- [Types](#types)
- [Multi-jobs](#multi-jobs-support)
- [Steps](#steps)
  * [goto](#goto) Navigate to an URL
  * [run](#run) Runs a group of steps by its name.
  * [goBack](#goBack) Navigate to the previous page in history
  * [screenshot](#screenshot) Takes an screenshot of the page
  * [pdf](#pdf) Takes a pdf of the page
  * [text](#text) Gets the text for a given CSS selector
  * [many](#many) Returns an array of elements given their CSS selectors
  * [title](#title) Gets the title for the current page.
  * [form](#form) Fill and submit forms
  * [html](#html) Return HTML code for the page or a DOM element
  * [innerHtml](#innerHtml) Return the 'visible' HTML code for a dom element
  * [click](#click) Click on an element (CSS and xPath selectors)
  * [url](#url) Return the current URL
  * [type](#type) Types a text (key events) in a given selector
  * [waitFor](#waitFor) Wait for selectors, time, functions, etc before continuing
  * [keyboardPress](#keyboardPress) Simulates the press of a keyboard key

## Overview

You can use WebParsy either as cli from your terminal or as a NodeJS library.

### Cli

*Install webparsy:*
```bash
$ npm i webparsy -g
```

```bash
$ webparsy example/_weather.yml --customFlag "custom flag value"
Result:

{
  "title": "Madrid, España Pronóstico del tiempo y condiciones meteorológicas - The Weather Channel | Weather.com",
  "city": "Madrid, España",
  "temp": 18
}
```

### Library

```javascript
const webparsy = require('webparsy')
const parsingResult = await webparsy.init({
  file: 'jobdefinition.yml'
  flags: { ... } // optional
})
```

#### Methods

##### init(options)

options:

One of `yaml`, `file` or `string` is required.

- `yaml`: A [yaml npm module](https://www.npmjs.com/package/yaml) instance of the scraping definition.
- `string`: The YAML definition, as a plain string.
- `file`: The path for the YAML file containing the scraping definition.

Additionally, you can pass a `flags` object property to input additional values
to your scraping process.

## Browser config

You can setup Chrome's details in the `browser` property within the main job.

None of the following settings are required.

```yaml
jobs:
  main:
    browser:
      width: 1200
      height: 800
      scaleFactor: 1
      timeout: 60
      delay: 0
      headless: true
      executablePath: ''
      userDataDir: ''
      keepOpen: false
```

- executablePath: If provided, webparsy will launch Chrome from the specified
path.
- userDataDir: If provided, webparsy will launch Chrome with the specified
user's profile.

## Output

In order for WebParsy to get contents, it needs some very basic details. This are:

- `as` the property you want to be returned
- `selector` the css selector to extract the html or text from

Other optional options are

- `parent` Get the parent of the element filtered by a selector. 

Example

```yaml
text:
  selector: .entry-title
  as: entryLink
  parent: a
```

## Transform

When you extract texts from a web page, you might want to transform the data
before returning them. [example](https://github.com/joseconstela/webparsy/blob/master/examples/methods/transform.yml)

You can use the following `- transform` methods:

- `uppercase` transforms the result to uppercase
- `lowercase` transforms the result to lowercase
- `absoluteUrl` return the absolute url for a link

## Types

When extractring details from a page, you might want them to be returned in
different formats, for example as a number in the example of grabing temperatures.
[example](https://github.com/joseconstela/webparsy/blob/master/examples/methods/type.yml)

You can use the following values for `- type`:

- `string`
- `number` 
- `integer` 
- `float` 
- `fcd` tranform to **f**loat an string-number that uses **c**omma for thousands
- `fdc` tranform to **f**loat an string-number that uses **d**ot for thousands

## Multi-jobs support

You can define groups of steps (jobs) that you can reuse at any moment during an
scraping process.

For example, let's say you want to signup twice in a website. You will have a
"main" job (that executes by defaul) but you can create an additional one called
"signup", that you can reuse in the "main" one.

```yaml
version: 1
jobs:
  main:
    steps:
      - goto: https://example.com/
      - run: signup
      - click: '#logout'
      - run: signup
  signup:
    steps:
      - goto: https://example.com/register
      - form:
          selector: "#signup-user"
          submit: true
          fill:
            - selector: '[name="username"]'
              value: jonsnow@example.com
```

## Steps

Steps are the list of things the browser must do.

This can be:

  * [goto](#goto) Navigate to an URL
  * [run](#run) Runs a group of steps by its name.
  * [goBack](#goBack) Navigate to the previous page in history
  * [screenshot](#screenshot) Takes an screenshot of the page
  * [pdf](#pdf) Takes a pdf of the page
  * [text](#text) Gets the text for a given CSS selector
  * [many](#many) Returns an array of elements given their CSS selectors
  * [title](#title) Gets the title for the current page.
  * [form](#form) Fill and submit forms
  * [html](#html) Return HTML code for the page or a DOM element
  * [innerHtml](#innerHtml) Return the 'visible' HTML code for a dom element
  * [click](#click) Click on an element (CSS and xPath selectors)
  * [url](#url) Return the current URL
  * [type](#type) Types a text (key events) in a given selector
  * [waitFor](#waitFor) Wait for selectors, time, functions, etc before continuing
  * [keyboardPress](#keyboardPress) Simulates the press of a keyboard key

## goto

URL to navigate page to. The url should include scheme, e.g. https://. [example](https://github.com/joseconstela/webparsy/blob/master/examples/methods/goBack.yml)

```yaml
- goto: https://example.com
```

You can also tell WebParsy to don't use Puppeteer to browse, and instead do a
direct HTTP request via got. This will perform much faster, but it may not be
suitable for websites that requires JavaScript. [simple example](https://github.com/joseconstela/webparsy/blob/master/examples/methods/getRequest.yml) / 
[extended example](https://github.com/joseconstela/webparsy/blob/master/examples/methods/many_using_get.yml)

Note that some methods that allows you to interact with a website (for example:
`form`, `click` and others) will not be be available if you specify to visit a
url via `got`:

```yaml
- goto:
    url: https://google.com
    method: got
```

You can tell WebParsy which urls it should visit via flags. For example for the
following definition:

```yaml
- goto:
    flag: websiteUrl
```

you can then specify the URL to be visited as

```bash
webparsy definition.yaml --websiteUrl "https://google.com"
```

or via the javascript library as:

```javascript
webparsy.init({
  file: 'definition.yml'
  flags: { websiteUrl: 'https://google.com' }
})
```

[example](https://github.com/joseconstela/webparsy/blob/master/examples/flags.js)

### Authentication

You can perform basic HTTP authentication by providing the user and password as
in the following example:

```yml
- goto: 
    url: http://example.com
    method: got
    authentication:
      type: basic
      username: my_user
      password: my_password
```

The authentication mechanism works for both puppeteer and `got` scrapping.

## run

Runs a group of steps by its name.

```yaml
- run: signupProcess
```

## goBack

Navigate to the previous page in history. [example](https://github.com/joseconstela/webparsy/blob/master/examples/methods/goBack.yml)

```yaml
- goBack
```

## screenshot

Takes an screenshot of the page. This triggers pupetteer's [page.screenshot](https://github.com/GoogleChrome/puppeteer/blob/v1.13.0/docs/api.md#pagescreenshotoptions).
[example](https://github.com/joseconstela/webparsy/blob/master/examples/methods/screenshot.yml)

```yaml
- screenshot:
  - path: Github.png
```

If you are using WebParsy as a NodeJS module, you can also get the screenshot
retuned as a Buffer by using the `as` property.

```yaml
- screenshot:
  - as: myScreenshotBuffer
```

## pdf

Takes a pdf of the page. This triggers pupetteer's [page.pdf](https://github.com/GoogleChrome/puppeteer/blob/v1.13.0/docs/api.md#pagepdfoptions)

```yaml
- pdf:
  - path: Github.pdf
```

If you are using WebParsy as a NodeJS module, you can also get the PDF file
retuned as a Buffer by using the `as` property.

```yaml
- pdf:
  - as: pdfFileBuffer
```

## title

Gets the title for the current page. If no output.as property is defined, the
page's title will tbe returned as `{ title }`. [example](https://github.com/joseconstela/webparsy/blob/master/examples/methods/goBack.yml)

```yaml
- title
```

## many

Returns an array of elements given their CSS selectors. [example](https://github.com/joseconstela/webparsy/blob/master/examples/methods/many.yml)

Example: 

```yaml
- many: 
  as: articles
  selector: main ol.articles-list li.article-item
  element:
    - text:
      selector: .title
      as: title
```

When you scape large amount of contents, you might end consuming hords of RAM,
your system might become slow and the scraping process might fail.

To prevent this, WebParsy allows you to use process events so you can have the
scraped contents as they are scraped, instead of storing them in memory and
waiting for the whole process to finish.

To do this, simply add an `event` property to `many`, with the event's name you
want to listen to. The event will contain each scraped item.

`event` will give you the data as it's being scraped. To prevent it from being
stored in memory, set `eventMethod` to `discard`.

[Example using events](https://github.com/joseconstela/webparsy/blob/master/examples/many_event.js)

## form

Fill and submit forms. [example](https://github.com/joseconstela/webparsy/blob/master/examples/methods/form.yml)

Form filling can use values from environment variables. This is useful if you
want to keep users login details in secret. If this is your case, instead of
specifying the value as a string, set it as the env property for value. Check
the example below or refer to [banking example](https://github.com/joseconstela/webparsy/blob/master/examples/methods/form.yml)

Example: 

```yaml
- form:
    selector: "#tsf"            # form selector
    submit: true               # Submit after filling all details
    fill:                      # array of inputs to fill
      - selector: '[name="q"]' # input selector
        value: test            # input value
```

Using environment variables
```yaml
- form:
    selector: "#login"            # form selector
    submit: true                  # Submit after filling all details
    fill:                         # array of inputs to fill
      - selector: '[name="user"]' # input selector
        value:
          env: USERNAME           # process.env.USERNAME
      - selector: '[name="pass"]' 
        value: 
          env: PASSWORD           # process.env.PASSWORD
```

## html

Gets the HTML code. If no `selector` specified, it returns the page's full HTML
code. If no output.as property is defined, the result will be returned
as `{ html }`. [example](https://github.com/joseconstela/webparsy/blob/master/examples/methods/html.yml)

Example: 

```yaml
- html
    as: divHtml
    selector: div
```

## innerHtml

Gets the innet HTML code for an element.
[example](https://github.com/joseconstela/webparsy/blob/master/examples/methods/innerHtml.yml)

This is useful when the javascript code of a page has changed the contents of
DOM, and you need to grab the actual HTML that a normal user would be
interacting with, instead the source code's html.

Example:

```yaml
- innerHtml
    as: divInnerHtml
    selector: div
```

## click

Click on an element. [example](https://github.com/joseconstela/webparsy/blob/master/examples/methods/click.yml)

Example:

Default behaviour (CSS selector)
```yaml
- click: button.click-me
```

Same as
```yaml
- click: 
    selector: button.click-me
```

By xPath (clicks on the first match)
```yaml
- click: 
    xPath: '/html/body/div[2]/div/div/div/div/div[3]/span'
```

## type

Sends a `keydown`, `keypress/input`, and `keyup` event for each character in
the text.

Example:

```yaml
- type:
    selector: input.user
    text: jonsnow@example.com
    options:
      delay: 4000
```

## url

Return the current URL.

Example:

```yaml
- url:
    as: currentUrl
```

## waitFor

Wait for specified CSS, XPath selectors, on an specific amount of time before
continuing [example](https://github.com/joseconstela/webparsy/blob/master/examples/methods/form.yml)

Examples: 

```yaml
- waitFor:
   selector: "#search-results"
```

```yaml
- waitFor:
   xPath: "/html/body/div[1]/header/div[1]/a/svg"
```

```yaml
- waitFor:
   function: "console.log(Date.now())"
```

```yaml
- waitFor:
    time: 1000 # Time in milliseconds
```

## keyboardPress

Simulates the press of a keyboard key. [extended docs](https://github.com/puppeteer/puppeteer/blob/master/docs/api.md#keyboardpresskey-options)

```yaml
- keyboardPress: 
    key: 'Enter'
```

## License

MIT © [Jose Constela](https://joseconstela.com)


CHANGELOG.md
# Changelog

## 0.8.0

- Add "keep open" support for development environments.
- Add "multi-jobs" support.
- Add keyboardPress support.
- Add waitForNavigation support.

## 0.7.7

- Force mocha exit

## 0.7.6

- Implement typing
- Implement custom chrome path and user data directory (Chrome profiles)

## 0.7.5

- Fix click by xpath

## 0.7.4

- Added support for .click with xPath
- Added support for .waitFor (xPath and function)


index.js
const debug = require('debug')('webparsy')

const definition = require('./helpers/definition')
const steps = require('./helpers/steps')
const fatal = require('./helpers/err').fatal

// Host a copy of the definition file parsed using yaml's module
let def = null

/**
 * Initialize the webparsy!
 * 
 * @param {Object} program Commander's original object
 */
const init = async (program) => {
  let error
  let exitCode = 0

  let output = {}
  // Get definition

  // File path specified from the cli
  if (program.file) {
    def = await definition.loadFile(program.file)
  }

  // YAML string
  else if (program.string) {
    def = await definition.loadString(program.string)
  }

  // YAML
  else if (program.yaml) {
    def = program.yaml
  }


  // No definition
  else {
    fatal('No definition specified')
  }

  await definition.validate(def)

  const browserOpts = definition.cfg(def, 'main', 'browser') || {}

  let options = {
    args: [
      '--no-sandbox'
    ],
    width: browserOpts.width || 1200,
    height: browserOpts.height || 800,
    scaleFactor: browserOpts.scaleFactor || 1,
    timeout: browserOpts.timeout || 30 * 1000,
    delay: browserOpts.delay || 0,
    headless: browserOpts.headless !== false
  }

  let puppeteer
  if (browserOpts.executablePath) {
    puppeteer = require('puppeteer-core')
    options.executablePath = browserOpts.executablePath
    options.userDataDir = browserOpts.userDataDir
  }
  else {
    puppeteer = require('puppeteer')
  }

  debug(`Puppeteer options ${JSON.stringify(options)}`)

  const browser = await puppeteer.launch(options)
  debug('Browser launched')

  const page = await browser.newPage()
  debug('New page created')

  const defaultSteps = definition.cfg(def, 'main', 'steps')
  debug('Get definition steps')
  try {
    const runSteps = async (allSteps, taskFlags) => {
      for (singleStep of allSteps) {
        if (singleStep.run) {
          const subTask = definition.cfg(def, singleStep.run)
          const subTaskSteps = subTask.steps
          const subTaskLoop = subTask.loop
          if (subTaskLoop) {
            if (!program.flags[subTaskLoop]) throw new Error('loop-undefined')
            for (item of program.flags[subTaskLoop]) {
              console.log(item)
              await runSteps(subTaskSteps, item)
              continue;
            }
          }
          else {
            await runSteps(subTaskSteps)
            continue;
          }
        }

        let stepResult = await steps.exec(taskFlags || program.flags, singleStep, page)
        if (stepResult.result) {
          output = Object.assign({}, output, stepResult.result || {})
        }
        else if (stepResult.flag) {
          program.flags = Object.assign({}, program.flags, stepResult.flag)
        }
      }
    }
    await runSteps(defaultSteps)
  }
  catch (ex) {
    error = ex
    exitCode = 5
  }
  finally { // Always gracefylly close the browser
    if (!browserOpts.keepOpen) await browser.close()
  }

  if (exitCode) {
    fatal(error)
  }

  if (program.cli) {
    console.log(JSON.stringify(output, ' ', 2))
  }
  else {
    return output
  }
}

module.exports.init = init